mlir-sycl/lib/Dialect/SYCL/IR/SYCLHostOps.cpp

```diff
@@ -10,3 +10,18 @@ LogicalResult SYCLHostConstructorOp::verify() {
            << type;
   return success();
 }
+
+void SYCLHostConstructorOp::getEffects(
+    SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
+        &effects) {
+  // NOTE: This definition conservatively assumes that all (pointer) arguments
+  // are written to. This is definitely true for the first argument ('this'
+  // pointer), but a pessimistic assumption for the remaining arguments.
+  auto *defaultResource = SideEffects::DefaultResource::get();
+  // The `this` argument will always be written to
+  effects.emplace_back(MemoryEffects::Write::get(), getDst(), defaultResource);
+  // For the remaining non-scalar arguments we also assume they are written.
+  for (auto value : getArgs())
+    if (isa<MemRefType, LLVM::LLVMPointerType>(value.getType()))
+      effects.emplace_back(MemoryEffects::Write::get(), value, defaultResource);
```

Comment:
Can we also add `Read` here? Also, what do you think of setting `Write` only for `accessor`?

Reply:
Added the `Read` effect. I would suggest we try how far the current, conservative effect definition gets us in terms of analysis and make the effects more optimistic by removing `Write` if necessary.